### DISCUSSION TITLE
Stubing a class during visiting a page.

### DISCUSSION THREAD

**Discussion by Philip-Grozdanov** [cypress-io/cypress#30428] [Questions and Help]
I want to stub the return value of a static method of a class(MyCustomClass.ts) after it gets loaded for the first time from visiting a page via ```cy.visit(url)```

My main page is "index.html". When I  visit this page all the typescript files get converted to .js files and then get loaded in the app via GET requests, an example request :  "GET", "http://127.0.0.1:9002/services/managers/MyCustomClass.js".
After all the files get loaded fila GET requests the backend calls a class called MainManager.ts (an entry point of the application) which after a number of calls invokes MyCustomClass.myMethod(), my goal
is to stub in that moment the return value of MyCustomClass.myMethod() to be "123" instead of the original one.

From code perspective what happens is : 
1.I visit the "index.html"
2.onBeforeLoad runs and stubs the static method
3.all the files load and MainManager gets invoked which prints the original value of MainManager.myMethod() instead of the stubbed "123"

```
//my-test.cy.ts
describe("set work schedule", () => {

  beforeEach(() => {
    mainPage.open();
  });
  
 
//MainPage.ts 
  open(): MainPage {
    clock.execAsOn(() => {
 
    cy.visit("index.html", {
      onBeforeLoad(win) {
        console.log("Load from OnBeforeLoad");
        win.MyCustomClass = MyCustomClass;
        console.log("Set ID:");
        win.customWindowId = "mainpageID"; //custom ID i have created to verify the window is the same
        console.log("window ID:" + win.customWindowId); //returns "mainpageID"
        const myMethodStub = cy.stub(win.MyCustomClass, "myMethod").returns("123");
      },
    });
 
 
      });
```

```
// MyCustomClass.ts
class MyCustomClass {
  // Define a static method
  public static myMethod(): string {
    return "MyCustomClass myMethod value";
  }

  // Define an instance method
  public anotherMethod(): string {
    return "MyCustomClass anotherMethod value";
  }
}
console.log("IN_CLASS_IMPL:");
if (window.MyCustomClass) console.log(window.MyCustomClass.toString()); //always returns  
console.log("window ID:" + window.customWindowId);

export default MyCustomClass;


```

``` cypress setup

import MyCustomClass from "ts/data/dao/MyCustomClass";

export {};
declare global {
  // Extend the Window interface in Cypress to include MyCustomClass
  interface Window {
    MyCustomClass: typeof MyCustomClass;
    customWindowId?: string;
  }
}


```

Observations : 

This is the order of events from the console:
![image](https://github.com/user-attachments/assets/6d9f09f0-7cb1-4cf2-a3e4-fd798d8218de)

1.Everwhere in the logs  win.customWindowId or window.MyCustomClass returns  "mainpageID"
2.From the MainManager.ts the lines which get printed : "===inside init===" and  "MyCustomClass myMethod value" , show that
the value is not stubbed  and also on the right invoking target is VM198... not a plain class.

Any suggestions how can I achieve this?

**Comment by portyu9** [comment]
Hi @Philip-Grozdanov 

The important detail is that `cy.stub()` only patches the exact object reference you give it.

This:

```ts
win.MyCustomClass = MyCustomClass
cy.stub(win.MyCustomClass, 'myMethod').returns('123')
```

creates/stubs `window.MyCustomClass`, but `MainManager.ts` is calling its own ES-module import:

```ts
import MyCustomClass from 'ts/data/dao/MyCustomClass'
```

Assigning something to `window.MyCustomClass` does not replace that import binding. The fact that `customWindowId` proves it is the same browser window does not make those two module references the same object.

For this kind of dependency, the reliable options are:

- inject the class/function into `MainManager` (constructor/factory/provider) so a test can supply a fake,
- deliberately make the application resolve the dependency from a test hook/global before startup, or
- if the method ultimately talks to a backend, stub the observable network boundary with `cy.intercept()` instead.

Trying to replace an internal imported module from an E2E spec is brittle because the AUT and the Cypress spec have separate module graphs.

Please mark this answer as accepted if it helped, thank you.
